### PR TITLE
Meta Tx

### PR DESCRIPTION
**Pull request by nick**
- My Listings / Sales / Purchases etc will show both wallet and proxy items
- Auto-deploy proxy the first time a user updates their identity / creates a listing / makes an offer

### CODE REVIEW

**Comment by nick** [comment]
Ready for another round of meta-tx to be merged in...

File: packages/graphql/src/utils/proxy.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  const { web3, ProxyImp, ProxyFactory } = contracts
+  let code = await ProxyFactory.methods.proxyCreationCode().call()
+  code += web3.eth.abi.encodeParameter('uint256', ProxyImp._address).slice(2)
+  return code
+})
+
+async function predictedProxyRaw(address) {
+  const web3 = contracts.web3Exec
+
+  // Calculate the call data used when deploying the contract
```
Comment (franckc) [inline packages/graphql/src/utils/proxy.js]: nit: cleanup if not used

File: packages/graphql/src/mutations/_txHelper.js [hunk 2]
```diff
@@ -34,13 +41,25 @@ function useRelayer({ mutation, value }) {
...
-  if (addr === proxy && mutation !== 'deployIdentityViaProxy') return false
+async function useProxy({ proxy, addr, to, from, mutation }) {
+  const { proxyAccountsEnabled } = contracts.config
+  if (!proxyAccountsEnabled) return false
   if (to) return false
   if (mutation === 'updateTokenAllowance') return false
   if (mutation === 'swapToToken') return false
+  if (!proxy && mutation === 'deployIdentity') return 'create'
+  if (!proxy && mutation === 'createListing') return 'create'
+  if (!proxy && mutation === 'makeOffer') {
```
[the remark this answers is not in the document]
Comment (nick) [reply packages/graphql/src/mutations/_txHelper.js]: This is in the useProxy method which is separate from relayer method... so yea we let them pay gas for it

File: packages/graphql/src/mutations/_txHelper.js [hunk 3]
```diff
@@ -145,6 +191,9 @@ export default function txHelper({
         })
       })
       .once('receipt', receipt => {
+        if (String(shouldUseProxy).indexOf('create') === 0) {
```
[the remark this answers is not in the document]
Comment (nick) [reply packages/graphql/src/mutations/_txHelper.js]: 👍

File: packages/graphql/src/utils/proxy.js (hunk 1 above)
Comment (nick) [reply packages/graphql/src/utils/proxy.js]: 👍

File: packages/contracts/contracts/identity/IdentityProxy.s [hunk 4]
```diff
@@ -227,4 +227,32 @@ contract IdentityProxy is ERC725 {
...
+        bytes _finalize,
+        address _seller,
+        address _currency,
+        uint _value
+    )
+        public
+        payable
+        onlyOwner
+    {
+        require(executeCall(_marketplace, 0, _finalize), 'marketplace-call-failed');
```
[the remark this answers is not in the document]
Comment (nick) [reply packages/contracts/contracts/identity/IdentityProxy.s]: Thanks, done

File: packages/graphql/src/mutations/_txHelper.js [hunk 5]
```diff
@@ -34,13 +41,25 @@ function useRelayer({ mutation, value }) {
   return window.localStorage.enableRelayer ? true : false
 }
 
-function useProxy({ proxy, addr, to, mutation }) {
-  if (!contracts.config.proxyAccountsEnabled) return false
-  if (!proxy) return false
-  if (addr === proxy && mutation !== 'deployIdentityViaProxy') return false
+async function useProxy({ proxy, addr, to, from, mutation }) {
```
[the remark this answers is not in the document]
Comment (nick) [reply packages/graphql/src/mutations/_txHelper.js]: Done

**Comment by nick** [comment]
Addressed @franckc comments and pushed some more stuff :-)

Ready for review 3 now... 😬

**Pull request merged**